pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go

```diff
@@ -0,0 +1,150 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package entrypoint
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/knative/build/pkg/apis/build/v1alpha1"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// MountName is the name of the pvc being mounted (which
+	// will contain the entrypoint binary and eventually the logs)
+	MountName        = "tools"
+	MountPoint       = "/tools"
+	BinaryLocation   = MountPoint + "/entrypoint"
+	JSONConfigEnvVar = "ENTRYPOINT_OPTIONS"
+	Image            = "gcr.io/k8s-prow/entrypoint@sha256:7c7cd8906ce4982ffee326218e9fc75da2d4896d53cabc9833b9cc8d2d6b2b8f"
+	InitContainerName = "place-tools"
+	ProcessLogFile = "/tools/process-log.txt"
+	MarkerFile = "/tools/marker-file.txt"
+)
+
+var toolsMount = corev1.VolumeMount{
+	Name:      MountName,
+	MountPath: MountPoint,
+}
+
+type Cache struct {
+	mtx sync.RWMutex
+	cache map[string][]string
+}
+
+func NewCache() *Cache {
+	return &Cache{
+		cache: make(map[string][]string),
+	}
+}
+
+func (c *Cache) get(sha string) ([]string, bool) {
+	c.mtx.RLock()
+	ep, ok := c.cache[sha]
+	c.mtx.RUnlock()
+	return ep, ok
+}
+
+func (c *Cache) set(sha string, ep []string) {
+	c.mtx.Lock()
+	c.cache[sha] = ep
+	c.mtx.Unlock()
+}
+
+// AddCopyStep will prepend a BuildStep (Container) that will
+// copy the entrypoint binary from the entrypoint image into the
+// volume mounted at MountPoint, so that it can be mounted by
+// subsequent steps and used to capture logs.
+func AddCopyStep(b *v1alpha1.BuildSpec) {
+	cp := corev1.Container{
+		Name:         InitContainerName,
+		Image:        Image,
+		Command:      []string{"/bin/cp"},
+		Args:         []string{"/entrypoint", BinaryLocation},
+		VolumeMounts: []corev1.VolumeMount{toolsMount},
+	}
+	b.Steps = append([]corev1.Container{cp}, b.Steps...)
+
+}
+
+type entrypointArgs struct {
+	Args       []string `json:"args"`
+	ProcessLog string   `json:"process_log"`
+	MarkerFile string   `json:"marker_file"`
+}
+
+func getEnvVar(cmd, args []string) (string, error) {
+	entrypointArgs := entrypointArgs{
+		Args:       append(cmd, args...),
+		ProcessLog: ProcessLogFile,
+		MarkerFile: MarkerFile,
+	}
+	j, err := json.Marshal(entrypointArgs)
+	if err != nil {
+		return "", fmt.Errorf("couldn't marshal arguments %q for entrypoint env var: %s", entrypointArgs, err)
+	}
+	return string(j), nil
+}
+
+func GetRemoteEntrypoint(cache *Cache, image string) ([]string, error) {
```

Comment:
you had really great comments on the other functions! maybe one here too? :D

Reply:
🤦‍♂️

Reply:
thanks!! :heart: